terminal/plugins/org.eclipse.tm.terminal.view.ui/src/org/eclipse/tm/terminal/view/ui/tabs/TabTerminalListener.java

```diff
@@ -161,10 +163,38 @@ protected String getTerminalConsoleTabTitle(TerminalState state) {
 		return newTitle != null && !newTitle.equals(oldTitle) ? newTitle : null;
 	}
 
-	@Override
-	public void setTerminalTitle(String title) {
-		tabItemTitle = title;
-		updateTitle();
+	/**
+	 * Sets Terminal tilte and checks if originator is ANSI command.
+	 * If originator is ANSI command in terminal and user does not want to use
+	 * ANSI command to upate terminal then return else update title.
+	 * @param title Title to update.
+	 * @param requestor Who requests title update. Can be null.
+	 * @since 11.3
+	 */
+	public void setTerminalTitle(String title, String requestor) {
+
+		CTabItem item = getTabItem();
+		if (item == null || item.isDisposed()) {
+			return;
+		}
+
+		// Run asynchronously in the display thread
+		item.getDisplay().asyncExec(() -> {
+			Boolean flag = false;
+			// Get the original terminal properties associated with the tab item
+			Map<String, Object> properties = (Map<String, Object>) item.getData("properties"); //$NON-NLS-1$
+			if (properties.containsKey(ITerminalsConnectorConstants.PROP_TITLE_UPDATE_API)) {
+				flag = (Boolean) properties.get(ITerminalsConnectorConstants.PROP_TITLE_UPDATE_API);
+			}
+			// Check if terminal should be updated either using API only (flag == true)
+			// or using API and ANSI command (flag == false).
+			if (flag == true && requestor != null && requestor.equals("ANSI")) {
+				return;
+			}
```

Comment:
This looks mostly like the inlined version of `updateTitle`. You can refactor private methods so we don't need to duplicate this code.